corehq/apps/case_search/utils.py

```diff
@@ -400,12 +399,38 @@ def get_related_cases_result(helper, app, case_types, source_cases, include_all_
         return _get_all_related_cases(helper, source_cases)
     else:
         results = []
-        results.extend(_get_search_detail_path_defined_cases(helper, app, case_types, source_cases))
-        source_case_ids = {case.case_id for case in source_cases}
-        results.extend(_get_child_cases_referenced_in_app(helper, app, case_types, source_case_ids))
+        with helper.profiler.timing_context('get_app_context'):
+            paths, child_case_types = get_app_context(helper.domain, app_id, case_types)
+        if paths:
+            results.extend(get_path_related_cases_results(helper, source_cases, paths))
+        if child_case_types:
+            source_case_ids = {case.case_id for case in source_cases}
+            results.extend(get_child_case_results(helper, source_case_ids, child_case_types))
         return results
 
 
+def get_app_context(domain, app_id, case_types):
+    """
+    for the provided case types, return paths to required case types, plus child case types
+    """
+    all_paths, all_child_case_types = _get_app_context_by_case_type(domain, app_id)
+    paths = set().union(*(all_paths[ct] for ct in case_types if ct in all_paths))
+    child_case_types = set().union(*(
+        all_child_case_types[ct] for ct in case_types if ct in all_child_case_types))
+    return paths, child_case_types
+
+
+@quickcache(['domain', 'app_id'], timeout=24 * 60 * 60)
+def _get_app_context_by_case_type(domain, app_id):
```

Comment:
I noticed today looking into `get_app_cached` that the cache that it manually uses is cleared elsewhere here https://github.com/dimagi/commcare-hq/blob/abd70322267467f28db83ff48171624b02116462/corehq/apps/app_manager/models.py#L4553-L4554. I haven't looked closely at the logic for clearing the cache there, but if it's good to bust that cache, it's probably also good to clear this cache there.

Reply:
Hmm, good find.  Looks like that was introduced here: https://github.com/dimagi/commcare-hq/pull/24664
My logic was that app builds don't change, so there's no need to expire the cache, but that points out that builds can be tagged as released or not (and I think was in the context around adding build profiles, which was further work of that nature).

However I do think this should probably mirror the logic of `get_app_cached`, especially to account for testing scenarios where `app_id` isn't a build.  I'll rework the caching here.